Orange/data/io.py

```diff
@@ -337,248 +349,171 @@ def open(filename, *args, **kwargs):
+                    return None
 
-        def is_number(item):
-            try: float(item)
-            except ValueError: return False
-            return True
         # Second row items are type identifiers
         def header_test2(items):
             return all(map(_RE_TYPES.match, items))
+
         # Third row items are flags and column attributes (attr=value)
         def header_test3(items):
             return all(map(_RE_FLAGS.match, items))
 
-        data = iter(data)
-        header_rows = []
-
-        # Try to parse a three-line header
-        lines = []
-        try:
-            lines.append(list(next(data)))
-            lines.append(list(next(data)))
-            lines.append(list(next(data)))
-        except StopIteration:
-            lines, data = [], chain(lines, data)
-        if lines:
-            l1, l2, l3 = lines
-            # Three-line header if line 2 & 3 match (1st line can be anything)
-            if header_test2(l2) and header_test3(l3):
-                header_rows = [l1, l2, l3]
-            else:
-                lines, data = [], chain((l1, l2, l3), data)
-
-        # Try to parse a single-line header
-        if not header_rows:
-            try: lines.append(list(next(data)))
-            except StopIteration: pass
-            if lines:
-                # Header if none of the values in line 1 parses as a number
-                if not all(is_number(i) for i in lines[0]):
-                    header_rows = [lines[0]]
-                else:
-                    data = chain(lines, data)
-
-        return header_rows, data
-
-    @classmethod
-    def data_table(self, data, headers=None):
-        """
-        Return Orange.data.Table given rows of `headers` (iterable of iterable)
-        and rows of `data` (iterable of iterable; if ``numpy.ndarray``, might
-        as well **have it sorted column-major**, e.g. ``order='F'``).
-
-        Basically, the idea of subclasses is to produce those two iterables,
-        however they might.
-
-        If `headers` is not provided, the header rows are extracted from `data`,
-        assuming they precede it.
-        """
-        if not headers:
-            headers, data = self.parse_headers(data)
-
-        # Consider various header types (single-row, two-row, three-row, none)
-        if 3 == len(headers):
-            names, types, flags = map(list, headers)
-        else:
-            if 1 == len(headers):
+        # read and parse the header before reading the rest of the file
+        header_df = self.read_header()
+        is_small = len(header_df) < 3
+
+        # the header can have 0, 1 or 3 columns
+        # is a three-row header if the second and third columns match,
+        # the first one can then be anything
+        three_row_header = not is_small and header_test2([str(i) for i in header_df.iloc[1].fillna('')]) \
+                           and header_test3([str(i) for i in header_df.iloc[2].fillna('')])
```

Comment:
Can also use `map()` here since it's used in header_test already. Can also use `.astype(str)`??